File: app/src/main/java/com/faceunity/fulivedemo/ui/control/AnimojiControlView.java
package com.faceunity.fulivedemo.ui.control;

import android.animation.ValueAnimator;
import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

import com.faceunity.OnFUControlListener;
import com.faceunity.entity.Effect;
import com.faceunity.fulivedemo.R;
import com.faceunity.fulivedemo.entity.CartoonFilter;
import com.faceunity.fulivedemo.entity.CartoonFilterEnum;
import com.faceunity.fulivedemo.entity.EffectEnum;
import com.faceunity.fulivedemo.ui.CheckGroup;
import com.faceunity.fulivedemo.ui.adapter.BaseRecyclerAdapter;

import java.util.List;

/**
 * Animoji 和动漫滤镜效果
 *
 * @author Richie on 2018.11.13
 */
public class AnimojiControlView extends FrameLayout implements CheckGroup.OnCheckedChangeListener {
    private static final int DEFAULT_FILTER_INDEX = 0;
    private static final int DEFAULT_ANIMOJI_INDEX = 0;
    private RecyclerView mRvAnim;
    private RecyclerView mRvFilter;
    private OnFUControlListener mOnFUControlListener;
    private boolean mIsShown;
    private CheckGroup mCheckGroup;
    private int mLastCheckedId = View.NO_ID;
    private OnBottomAnimatorChangeListener mOnBottomAnimatorChangeListener;
    private ValueAnimator mBottomLayoutAnimator;

    public AnimojiControlView(@NonNull Context context) {
        super(context);
        initView(context);
    }

    public AnimojiControlView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        initView(context);
    }

    public AnimojiControlView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(context);
    }

    private void initView(Context context) {
        View view = LayoutInflater.from(context).inflate(R.layout.layout_animoji_filter, this);
        mCheckGroup = view.findViewById(R.id.rg_anim);
        mCheckGroup.setOnCheckedChangeListener(this);
        mRvAnim = view.findViewById(R.id.rv_animoji);
        mRvAnim.setHasFixedSize(true);
        mRvAnim.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        ((SimpleItemAnimator) mRvAnim.getItemAnimator()).setSupportsChangeAnimations(false);
        AnimojiAdapter animojiAdapter = new AnimojiAdapter(EffectEnum.getEffectsByEffectType(Effect.EFFECT_TYPE_ANIMOJI));
        animojiAdapter.setOnItemClickListener(new OnAnimojiItemClickListener());
        animojiAdapter.setItemSelected(DEFAULT_ANIMOJI_INDEX);
        mRvAnim.setAdapter(animojiAdapter);
        mRvFilter = view.findViewById(R.id.rv_filter);
        mRvFilter.setHasFixedSize(true);
        mRvFilter.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        ((SimpleItemAnimator) mRvFilter.getItemAnimator()).setSupportsChangeAnimations(false);
        FilterAdapter filterAdapter = new FilterAdapter(CartoonFilterEnum.getAllCartoonFilters());
        filterAdapter.setOnItemClickListener(new OnFilterItemClickListener());
        filterAdapter.setItemSelected(DEFAULT_FILTER_INDEX);
        mRvFilter.setAdapter(filterAdapter);
        // 默认选中 Animoji 页面无选中，开启第一个动漫滤镜
        getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {

            @Override
            public void onGlobalLayout() {
                getViewTreeObserver().removeOnGlobalLayoutListener(this);
                onCheckedChanged(mCheckGroup, R.id.cb_animoji);
            }
        });
    }

    public void setOnFUControlListener(OnFUControlListener onFUControlListener) {
        mOnFUControlListener = onFUControlListener;
    }

    @Override
    public void onCheckedChanged(CheckGroup group, int checkedId) {
        mRvFilter.setVisibility(GONE);
        mRvAnim.setVisibility(GONE);
        if (checkedId == R.id.cb_animoji) {
            mRvAnim.setVisibility(VISIBLE);
        } else if (checkedId == R.id.cb_filter) {
            mRvFilter.setVisibility(VISIBLE);
        }
        if ((checkedId == View.NO_ID || checkedId == mLastCheckedId) && mLastCheckedId != View.NO_ID) {
            int endHeight = (int) getResources().getDimension(R.dimen.x98);
            int startHeight = getMeasuredHeight();
            changeBottomLayoutAnimator(startHeight, endHeight);
            mIsShown = false;
        } else if (checkedId != View.NO_ID && mLastCheckedId == View.NO_ID) {
            int startHeight = (int) getResources().getDimension(R.dimen.x98);
            int endHeight = (int) getResources().getDimension(R.dimen.x266);
            changeBottomLayoutAnimator(startHeight, endHeight);
            mIsShown = true;
        }
        mLastCheckedId = checkedId;
    }

    @Override
    public boolean isShown() {
        return mIsShown;
    }

    public void setOnBottomAnimatorChangeListener(OnBottomAnimatorChangeListener onBottomAnimatorChangeListener) {
        mOnBottomAnimatorChangeListener = onBottomAnimatorChangeListener;
    }

    public void hideBottomLayoutAnimator() {
        mCheckGroup.check(View.NO_ID);
    }

    private void changeBottomLayoutAnimator(final int startHeight, final int endHeight) {
        if (mBottomLayoutAnimator != null && mBottomLayoutAnimator.isRunning()) {
            mBottomLayoutAnimator.end();
        }
        mBottomLayoutAnimator = ValueAnimator.ofInt(startHeight, endHeight).setDuration(150);
        mBottomLayoutAnimator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                int height = (int) animation.getAnimatedValue();
                ViewGroup.LayoutParams params = getLayoutParams();
                params.height = height;
                setLayoutParams(params);
                if (mOnBottomAnimatorChangeListener != null) {
                    float showRate = 1.0f * (height - startHeight) / (endHeight - startHeight);
                    mOnBottomAnimatorChangeListener.onBottomAnimatorChangeListener(startHeight > endHeight ? 1 - showRate : showRate);
                }
            }
        });
        mBottomLayoutAnimator.start();
    }

    public interface OnBottomAnimatorChangeListener {
        void onBottomAnimatorChangeListener(float showRate);
    }

    private class OnFilterItemClickListener implements BaseRecyclerAdapter.OnItemClickListener<CartoonFilter> {

        @Override
        public void onItemClick(BaseRecyclerAdapter<CartoonFilter> adapter, View view, int position) {
            CartoonFilter cartoonFilter = adapter.getItem(position);
            if (mOnFUControlListener != null && cartoonFilter != null) {
                mOnFUControlListener.setCartoonFilter(cartoonFilter.getStyle());
            }
        }
    }

    private class OnAnimojiItemClickListener implements BaseRecyclerAdapter.OnItemClickListener<Effect> {

        @Override
        public void onItemClick(BaseRecyclerAdapter<Effect> adapter, View view, int position) {
            Effect effect = adapter.getItem(position);
            if (mOnFUControlListener != null) {
                mOnFUControlListener.onEffectSelected(effect);
            }
        }
    }

    private static class FilterAdapter extends BaseRecyclerAdapter<CartoonFilter> {

        public FilterAdapter(@NonNull List<CartoonFilter> data) {
            super(data, R.layout.layout_animoji_recycler);
        }

        @Override
        protected void bindViewHolder(BaseViewHolder viewHolder, CartoonFilter item) {
            viewHolder.setImageResource(R.id.iv_anim_item, item.getIconId());
        }

        @Override
        protected void handleSelectedState(BaseViewHolder viewHolder, CartoonFilter data, boolean selected) {
            viewHolder.setBackground(R.id.iv_anim_item, selected ? R.drawable.effect_select : 0);
        }
    }

    private static class AnimojiAdapter extends BaseRecyclerAdapter<Effect> {

        public AnimojiAdapter(@NonNull List<Effect> data) {
            super(data, R.layout.layout_animoji_recycler);
        }

        @Override
        protected void bindViewHolder(BaseViewHolder viewHolder, Effect item) {
            viewHolder.setImageResource(R.id.iv_anim_item, item.getIconId());
        }

        @Override
        protected void handleSelectedState(BaseViewHolder viewHolder, Effect data, boolean selected) {
            viewHolder.setBackground(R.id.iv_anim_item, selected ? R.drawable.effect_select : 0);
        }
    }

}

File: fu_ui/src/main/java/com/faceunity/ui/control/AnimojiControlView.kt
package com.faceunity.ui.control

import android.animation.ValueAnimator
import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.faceunity.ui.R
import com.faceunity.ui.base.BaseDelegate
import com.faceunity.ui.base.BaseListAdapter
import com.faceunity.ui.base.BaseViewHolder
import com.faceunity.ui.databinding.LayoutAnimoControlBinding
import com.faceunity.ui.entity.AnimationFilterBean
import com.faceunity.ui.entity.AnimojiBean
import com.faceunity.ui.infe.AbstractAnimojiDataFactory


/**
 *
 * DESC：
 * Created on 2020/12/10
 *
 */
class AnimojiControlView @JvmOverloads constructor(
    private val mContext: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) :
    BaseControlView(mContext, attrs, defStyleAttr) {


    private lateinit var mDataFactory: AbstractAnimojiDataFactory

    private lateinit var mAnimojiAdapter: BaseListAdapter<AnimojiBean>
    private lateinit var mAnimationFilterAdapter: BaseListAdapter<AnimationFilterBean>

    private val mBinding: LayoutAnimoControlBinding by lazy {
        LayoutAnimoControlBinding.inflate(LayoutInflater.from(context), this, true)
    }


    // region  init
    init {
        initView()
        initAdapter()
        bindListener()
    }

    /**
     * 给控制绑定AnimojiController，数据工厂
     */
    fun bindDataFactory(dataFactory: AbstractAnimojiDataFactory) {
        mDataFactory = dataFactory
        mAnimojiAdapter.setData(dataFactory.animojis)
        mAnimationFilterAdapter.setData(dataFactory.filters)
    }

    /**
     * 收回菜单栏
     */
    fun hideControlView() {
        mBinding.rgAnim.check(View.NO_ID)
    }


    private fun initView() {
        isBottomShow = true
        initHorizontalRecycleView(mBinding?.recyclerView)
    }

    private fun initAdapter() {
        mAnimojiAdapter = BaseListAdapter(ArrayList(), object : BaseDelegate<AnimojiBean>() {
            override fun convert(
                viewType: Int,
                helper: BaseViewHolder,
                data: AnimojiBean,
                position: Int
            ) {
                helper.setImageResource(R.id.iv_control, data.iconId)
                helper.itemView.isSelected = position == mDataFactory.currentAnimojiIndex
            }

            override fun onItemClickListener(view: View, data: AnimojiBean, position: Int) {
                if (mDataFactory.currentAnimojiIndex != position) {
                    changeAdapterSelected(
                        mAnimojiAdapter,
                        mDataFactory.currentAnimojiIndex,
                        position
                    )
                    mDataFactory.currentAnimojiIndex = position
                    mDataFactory.onAnimojiSelected(data)

                }
            }
        }, R.layout.list_item_control_image_circle)
        mBinding.recyclerView.adapter = mAnimojiAdapter

        mAnimationFilterAdapter =
            BaseListAdapter(ArrayList(), object : BaseDelegate<AnimationFilterBean>() {
                override fun convert(
                    viewType: Int,
                    helper: BaseViewHolder,
                    data: AnimationFilterBean,
                    position: Int
                ) {
                    helper.setImageResource(R.id.iv_control, data.iconId)
                    helper.itemView.isSelected = position == mDataFactory.currentFilterIndex
                }

                override fun onItemClickListener(
                    view: View,
                    data: AnimationFilterBean,
                    position: Int
                ) {
                    if (mDataFactory.currentFilterIndex != position) {
                        changeAdapterSelected(
                            mAnimationFilterAdapter,
                            mDataFactory.currentFilterIndex,
                            position
                        )
                        mDataFactory.currentFilterIndex = position
                        mDataFactory.onFilterSelected(data)
                    }
                }
            }, R.layout.list_item_control_image_circle)
    }

    private fun bindListener() {
        mBinding.rgAnim.setOnCheckedChangeListener { _, checkedId ->
            when (checkedId) {
                R.id.cb_animoji -> {
                    changeBottomLayoutAnimator(true)
                    mBinding.recyclerView.adapter = mAnimojiAdapter
                }

                R.id.cb_filter -> {
                    changeBottomLayoutAnimator(true)
                    mBinding.recyclerView.adapter = mAnimationFilterAdapter
                }

                else -> {
                    changeBottomLayoutAnimator(false)
                }
            }
        }
    }
    // endregion


    /**
     * 底部动画处理
     * @param isOpen Boolean
     */
    private fun changeBottomLayoutAnimator(isOpen: Boolean) {
        if (isOpen == isBottomShow) {
            return
        }

        val start =
            if (isOpen) resources.getDimensionPixelSize(R.dimen.x98) else resources.getDimensionPixelSize(
                R.dimen.x266
            )
        val end =
            if (isOpen) resources.getDimensionPixelSize(R.dimen.x266) else resources.getDimensionPixelSize(
                R.dimen.x98
            )

        if (bottomLayoutAnimator != null && bottomLayoutAnimator!!.isRunning) {
            bottomLayoutAnimator!!.end()
        }
        bottomLayoutAnimator = ValueAnimator.ofInt(start, end).setDuration(150)
        bottomLayoutAnimator!!.addUpdateListener { animation ->
            val height = animation.animatedValue as Int
            val params = mBinding.lytBottom.layoutParams as ViewGroup.LayoutParams
            params.height = height
            mBinding.lytBottom.layoutParams = params
            if (onBottomAnimatorChangeListener != null) {
                val showRate = 1.0f * (height - start) / (end - start)
                onBottomAnimatorChangeListener?.onBottomAnimatorChangeListener(if (!isOpen) 1 - showRate else showRate)
            }
        }
        bottomLayoutAnimator!!.start()
        isBottomShow = isOpen
    }

}